src/compiler/checker.ts

```diff
@@ -20521,23 +20536,28 @@ namespace ts {
             return type;
         }
 
+        function getReturnOrPromisedType(node: ArrowFunction | FunctionExpression | FunctionDeclaration | MethodDeclaration | MethodSignature, functionFlags: FunctionFlags) {
+            const returnTypeNode = getEffectiveReturnTypeNode(node);

```

Comment:
This is only used if `(functionFlags & FunctionFlags.AsyncGenerator) !== FunctionFlags.Async`, could it be computed only then?

Reply:
checkAsyncFunctionReturnType immediately calls getEffectiveReturnTypeNode, and assumes that its result is defined, so it implicitly used it. I added a parameter `returnTypeNode` to make the dependency explicit (and get rid of yet another TODO #18217)